src/compiler/utilities.ts

```diff
@@ -1756,6 +1796,26 @@ namespace ts {
         };
     }
 
+    export function makeModulePathSemiAbsolute(host: EmitHost, currentSourceFile: SourceFile, externalModuleName: string): string {
+        let quotationMark = externalModuleName.charAt(0);

```

Comment:
my only comment here, is we can simplify this using the resolver to get the symbol and then use the symbol's declaration filename to get the semiAbsolutePath,

here is what i had in mind.
```ts
        function emitExternalModuleSpecifier(moduleSpecifier: Expression) {
            Debug.assert(moduleSpecifier.kind === SyntaxKind.StringLiteral);

            if (compilerOptions.out) {
                let moduleSymbol = resolver.getSymbolAtLocation(moduleSpecifier);
                if (moduleSymbol && moduleSymbol.valueDeclaration &&
                    moduleSymbol.valueDeclaration.kind === SyntaxKind.SourceFile &&
                    !isDeclarationFile(<SourceFile>moduleSymbol.valueDeclaration)) {
                    let nonRelativeModuleName = getExternalModuleNameFromPath((<SourceFile>moduleSymbol.valueDeclaration).fileName);
                    write("\"");
                    write(nonRelativeModuleName);
                    write("\"");
                    return;
                }
            }

            writeTextOfNode(currentSourceFile, moduleSpecifier);
        }

        function getExternalModuleNameFromPath(fileName: string) {
            let sourceFilePath = getNormalizedAbsolutePath(fileName, host.getCurrentDirectory());
            let commonSourceDirectory = host.getCommonSourceDirectory();
            sourceFilePath = sourceFilePath.replace(commonSourceDirectory, "");
            return removeFileExtension(sourceFilePath);
        }
```